File: src/ClientModel/ArrayType.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;

namespace AutoRest.Java.Model
{
    /// <summary>
    /// The details of an array type that is used by a client.
    /// </summary>
    public class ArrayType : IType
    {
        public static readonly ArrayType ByteArray = new ArrayType(PrimitiveType.Byte, (string defaultValueExpression) => defaultValueExpression == null ? "new byte[0]" : $"\"{defaultValueExpression}\".getBytes()");

        private ArrayType(IType elementType, Func<string,string> defaultValueExpressionConverter)
        {
            ElementType = elementType;
            DefaultValueExpressionConverter = defaultValueExpressionConverter;
        }

        public IType ElementType { get; }

        private Func<string, string> DefaultValueExpressionConverter { get; }

        public override string ToString()
        {
            return $"{ElementType}[]";
        }
        
        public IType AsNullable()
        {
            return this;
        }

        public bool Contains(IType type)
        {
            return this == type || ElementType.Contains(type);
        }

        public void AddImportsTo(ISet<string> imports, bool includeImplementationImports)
        {
            ElementType.AddImportsTo(imports, includeImplementationImports);
        }

        public string DefaultValueExpression(string sourceExpression)
        {
            return DefaultValueExpressionConverter(sourceExpression);
        }

        public IType ClientType
        {
            get
            {
                // The only supported array type is byte[]
                return this;
            }
        }

        public string ConvertToClientType(string expression)
        {
            // The only supported array type is byte[]
            return expression;
        }

        public string ConvertFromClientType(string expression)
        {
            // The only supported array type is byte[]
            return expression;
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/models/clientmodel/ArrayType.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.azure.autorest.models.clientmodel;

import java.util.Set;
import java.util.function.Function;


/** 
 The details of an array type that is used by a client.
*/
public class ArrayType implements IType
{
	public static final ArrayType ByteArray = new ArrayType(PrimitiveType.Byte, (String defaultValueExpression) -> defaultValueExpression == null ? "new byte[0]" : String.format("\"%1$s\".getBytes()", defaultValueExpression));

	private ArrayType(IType elementType, Function<String,String> defaultValueExpressionConverter)
	{
		ElementType = elementType;
		DefaultValueExpressionConverter = (String arg) -> defaultValueExpressionConverter.apply(arg);
	}

	private IType ElementType;
	public final IType getElementType()
	{
		return ElementType;
	}

	private Function<String, String> DefaultValueExpressionConverter;
	private Function<String, String> getDefaultValueExpressionConverter()
	{
		return DefaultValueExpressionConverter;
	}

	@Override
	public String toString()
	{
		return String.format("%1$s[]", getElementType());
	}

	public final IType AsNullable()
	{
		return this;
	}

	public final boolean Contains(IType type)
	{
		return this == type || getElementType().Contains(type);
	}

	public final void AddImportsTo(Set<String> imports, boolean includeImplementationImports)
	{
		getElementType().AddImportsTo(imports, includeImplementationImports);
	}

	public final String DefaultValueExpression(String sourceExpression)
	{
		return DefaultValueExpressionConverter.apply(sourceExpression);
	}

	public final IType getClientType()
	{
			// The only supported array type is byte[]
		return this;
	}

	public final String ConvertToClientType(String expression)
	{
		// The only supported array type is byte[]
		return expression;
	}

	public final String ConvertFromClientType(String expression)
	{
		// The only supported array type is byte[]
		return expression;
	}
}